Backend for local simulator #322

lavanya-m-k opened the pull request:
# Description

Adding support for local quantum simulators with Qiskit Aer

Fixes #322

# Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] This change requires a documentation update

# How Has This Been Tested?

Added test cases. Tested by running `mgym dispatch metriq_gym/schemas/examples/bseq.example.json --provider local --device aer_simulator`

# Checklist:

- [x] I have performed a self-review of my code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules (or not applicable)

cosenal commented:
I tried to run the example you put in the PR description, and the simulator hangs forever, without yielding any result.

cosenal commented:
@lavanya-m-k Interesting. I can investigate why BSEQ hangs there. In theory it's not a heavy benchmark to run.
In the meantime, please take a look at my other comments. Thanks

lavanya-m-k commented on metriq_gym/helpers/graph_helpers.py:
```diff
@@ -65,9 +65,12 @@ def device_graph_coloring(topology_graph: rx.PyGraph) -> GraphColoring:
-    # Graphs are bipartite, so use that feature to prevent extra colors from greedy search.
-    # This graph is colored using a bipartite edge-coloring algorithm.
-    edge_color_map = rx.graph_bipartite_edge_color(topology_graph)
+    if rx.is_bipartite(topology_graph):
+        # Bipartite graphs admit an optimal coloring using this specialized algorithm.
+        edge_color_map = rx.graph_bipartite_edge_color(topology_graph)
+    else:
+        # Fall back to a greedy edge-coloring for non-bipartite graphs.
+        edge_color_map = rx.graph_greedy_edge_color(topology_graph)
```
I encountered a GraphNotBipartite error when running the bseq benchmark on the AerSimulator backend due to the assumption that all device graphs would be bipartite. To make the code more robust, I added a fallback using graph_greedy_edge_color when the bipartite assumption fails.

While this wasn't the original intention of the PR, it was necessary to ensure compatibility with the new local simulator integration. If preferred, I can move this into a separate PR.

lavanya-m-k commented on metriq_gym/local/aer.py:
```diff
@@ -0,0 +1,68 @@
+from qiskit_aer import AerSimulator
+
+
+LOCAL_JOB_DIR = ".metriq_local_jobs"
```
The main idea behind saving local jobs as files is to simulate a minimal job tracking mechanism similar to remote backends. This allows us to mimic result persistence and later retrieval, especially useful when integrating with workflows that expect a job_id and result-loading behavior.
That said, I’m happy to switch to an in-memory model or use a temporary file-based approach (like using Python's tempfile) if that better aligns with the project architecture. Please let me know your preference.

lavanya-m-k replied:
That makes sense. Changed back to same jsonl file

lavanya-m-k commented:
> @lavanya-m-k my bad, sorry, I typed the comments, but never submitted them 🤦 You can see they are 8 hours old, but weren't visible until now

Addressed the review comments. Could you please let me know if the changes looks good?

vprusso commented:
There appear to be some build failures, @lavanya-m-k 
https://github.com/unitaryfoundation/metriq-gym/actions/runs/15591468210/job/43911370074?pr=372

lavanya-m-k commented:
This is related to `rustworkx.GraphNotBipartite: Graph is not bipartite`. Itt goes away if I add the following lines in graph_helpers.py

```
 if rx.is_bipartite(topology_graph):
        # Bipartite graphs admit an optimal coloring using this specialized algorithm.
        edge_color_map = rx.graph_bipartite_edge_color(topology_graph)
    else:
        # Fall back to a greedy edge-coloring for non-bipartite graphs.
        edge_color_map = rx.graph_greedy_edge_color(topology_graph
```

> There appear to be some build failures, @lavanya-m-k https://github.com/unitaryfoundation/metriq-gym/actions/runs/15591468210/job/43911370074?pr=372

lavanya-m-k commented:
> There appear to be some build failures, @lavanya-m-k https://github.com/unitaryfoundation/metriq-gym/actions/runs/15591468210/job/43911370074?pr=372

Build should work now

vprusso commented on metriq_gym/local/aer.py:
```diff
@@ -0,0 +1,133 @@
+import os
+import uuid
+from datetime import datetime
+from typing import Optional, Union
```
[nit]: Can we please use native types for consistency? IOW, 

```suggestion
```

vprusso commented on metriq_gym/local/aer.py:
```diff
@@ -0,0 +1,133 @@
+        self.num_qubits = self.backend.configuration().n_qubits
+
+    def run(
+        self, circuits: Union[QuantumCircuit, list[QuantumCircuit]], shots: Optional[int] = None
```
```suggestion
        self, circuits: QuantumCircuit | list[QuantumCircuit], shots: int | None = None
```

vprusso commented on metriq_gym/local/aer.py:
```diff
@@ -0,0 +1,133 @@
+
+    def run(
+        self, circuits: Union[QuantumCircuit, list[QuantumCircuit]], shots: Optional[int] = None
+    ) -> Union[LocalJob, list[LocalJob]]:
```
```suggestion
    ) -> LocalJob | list[LocalJob]:
```

vprusso commented on metriq_gym/local/aer.py:
```diff
@@ -0,0 +1,133 @@
+            result = aer_job.result()
+            counts = result.get_counts(circ)
+            jobs.append(
+                LocalJob(
```
Do we want to be hardcoding specific benchmarks here for local jobs like quantum volume? Maybe I'm missing the overall intent of this block?

lavanya-m-k replied:
It was a dev test which was accidentally added. Now it is updated. Thanks for pointing it out

vprusso commented on tests/local/test_local_aer.py:
```diff
@@ -0,0 +1,55 @@
+import json
+import pytest
+
+pytest.importorskip("qiskit")
```
Is there a reason that we `importorskip`? My impression would be that we would likely require these to be imported if we are going to test them and to fail if they are not provided as a dependency.

lavanya-m-k replied:
Yes you are right. Updated it!

cosenal commented on metriq_gym/benchmarks/__init__.py:
```diff
@@ -16,6 +16,7 @@ class JobType(StrEnum):
     QUANTUM_VOLUME = "Quantum Volume"
     MIRROR_CIRCUITS = "Mirror Circuits"
     WORMHOLE = "Wormhole"
+    LOCAL = "Local"
```
Why is `Local` a job type?
The idea is that one "dispatches" a regular job type (e.g. "Wormhole") to a local simulator. What does dispatching a `Local` job type mean?

lavanya-m-k replied:
The intent is to distinguish between where we run a job versus what the job is (e.g. Wormhole, Quantum Volume). Right now `JobType` blurs those lines.

lavanya-m-k replied:
Next steps:

We can rename `JobType` to something like `JobCategory` or split it into two enums:

- `JobType` – illustrates what job we’re running (e.g., Quantum Volume, Mirror Circuits, Wormhole)

- `ExecutionTarget` – illustrates where it’s run (Local, Remote, IonQ, etc.)

cosenal commented on metriq_gym/job_manager.py:
```diff
@@ -21,6 +23,7 @@ class MetriqGymJob:
     provider_name: str
     device_name: str
     dispatch_time: datetime
+    result_data: list[dict[str, Any]] | None = None
```
I like the introduction of this. It could also be used for caching results from non-simulator runs.

lavanya-m-k replied:
That’s a great point — this wrapper isn’t just for simulation results. We could add a `source` or `origin` field (e.g., `simulator`, `quantum_device`, `cached_result`) to clarify the provenance of the result_data. This would make reuse and caching more explicit.

cosenal commented on metriq_gym/qplatform/device.py:
```diff
@@ -41,3 +50,9 @@ def _(device: BraketDevice) -> rx.PyGraph:
+
+
+@connectivity_graph.register
+def _(device: AerSimulatorDevice) -> rx.PyGraph: 
```
I would argue it's helpful to have it defined though. Let's say one wants to simulate BSEQ on a local simulator. With this change, that would work out of the box.
